### DISCUSSION TITLE
Does Rauc support OTA upgrade for cpio.gz compressed initramfs image ?

### DISCUSSION THREAD

**Discussion by saramahe** [rauc/rauc#1267] [Q&A]
Hi ,
We are using Raspberry PI CM4  for our embedded development. As per the requirements, We don't need permanent rootfs system in emmc, hence we are using initramfs image for our applications. Initramfs are loaded into  RAM and used as an initial root filesystem during the early boot process.

Our requirements for the OTA upgrade is below,
1.We need to have two partitions in emmc formatted to ext4. 
2.We need to have working application +kernel images stored in two partitons (.cpio.gz.u-boot + Image)
3.We need to load the images from one partiton , for example say /dev/mmcblk0p2, then we need to mark the partition as active and other one  /dev/mmcblk0p3 as inactive
4. Whenever there is an upgrade, we need to download the upgrade package bundle and need to extract to other inactive partition  /dev/mmcblk0p3 and then need to load the image from this partition to RAM. 

We are using u-boot as bootloader and using following commands,
load ${BOOT_DEV} ${kernel_addr_r} Image; 
load ${BOOT_DEV} ${ramdisk_addr_r}My-dev-image.rootfs.cpio.gz.u-boot;
booti ${kernel_addr_r} ${ramdisk_addr_r} ${fdt_addr}

We are using yocto as build environment. I tried to integrate meta-rauc to our system.

This is my sytem.conf,

```
[system]
compatible=RaspberryPi4
bootloader=uboot
mountprefix=/mnt/raucbundle

[keyring]
path=/etc/rauc/ca.cert.pem

[slot.rootfs.0]
device=/dev/mmcblk0p2
type=ext4
bootname=A

[slot.rootfs.1]
device=/dev/mmcblk0p3
type=ext4
bootname=B

[slot.kernel.0]
parent=rootfs.0
device=/dev/mmcblk0p2
type=ext4

[slot.kernel.1]
parent=rootfs.1
device=/dev/mmcblk0p3
type=ext4
```

This is bundle recipe,

```
DESCRIPTION = "RAUC bundle generator"
LIC_FILES_CHKSUM = "file://${COREBASE}/meta/COPYING.MIT;md5=3da9cfbcb788c80a0384361b4de20420"
inherit bundle

RAUC_BUNDLE_COMPATIBLE = "RaspberryPi4"
RAUC_BUNDLE_VERSION = "v20200703"
RAUC_BUNDLE_DESCRIPTION = "RAUC Demo Bundle"

SRC_URI += "file://Other_STM_0.0.0.bin"

RAUC_BUNDLE_FORMAT = "verity"

RAUC_BUNDLE_SLOTS = "rootfs kernel"
RAUC_SLOT_rootfs = "My-dev-image"
RAUC_SLOT_rootfs[rename] = "My-dev-image.rootfs.cpio.gz.u-boot"


RAUC_SLOT_kernel = "linux-raspberrypi"
RAUC_SLOT_kernel[type] = "file"
RAUC_SLOT_kernel[file] = "Image"


RAUC_BUNDLE_EXTRA_FILES += "Other_STM_0.0.0.bin"

RAUC_KEY_FILE = "${THISDIR}/files/development-1.key.pem"
RAUC_CERT_FILE = "${THISDIR}/files/development-1.cert.pem"
```

I got these errors in my target , when I try to install  the bundle.

```
root@mcu632-cm4-64:/etc# rauc status

(rauc:483): rauc-WARNING **: 12:38:26.457: Failed getting primary slot: Failed getting primary slot: Unable to find primary boot slot

=== System Info ===
Compatible:  RaspberryPi4
Variant:
Booted from: rootfs.0 (A)

=== Bootloader ===
Activated: (null) ((null))

=== Slot States ===
o [rootfs.1] (/dev/mmcblk0p3, ext4, inactive)
        bootname: B
        mounted: /media/config
        boot status: bad
    [kernel.1] (/dev/mmcblk0p3, ext4, inactive)

o [rootfs.0] (/dev/mmcblk0p2, ext4, booted)
        bootname: A
        boot status: bad
    [kernel.0] (/dev/mmcblk0p2, ext4, active)

root@mcu632-cm4-64:/media/config# rauc install update-bundle-raspberrypi4-64.raucb
installing
  0% Installing
  0% Determining slot states
 20% Determining slot states done.
 20% Checking bundle
 20% Verifying signature
 40% Verifying s[   56.855577] loop0: detected capacity change from 0 to 85291
ignature done.
 40% Checking bundle done.
[   56.889542] loop0: detected capacity change from 85291 to 85288
[   56.916127] device-mapper: verity: sha256 using implementation "sha256-generic"
 40% Checking manifest contents
 60% Checking manifest contents done.
 60% Determining target install group
 80% Determining target install group done.
 80% Updating slots
 99% Updating slots failed.
100% Installing failed.
LastError: Installation error: Unsupported image /mnt/raucbundle/bundle/My-dev-image.cpio.gz.u-boot for slot type ext4
Installing `/media/config/update-bundle-raspberrypi4-64.raucb` failed
```
 I did check this cause of this error in the rauc code, I could see there is no default handler for this image  and slot type. 

how to do installation with this image type?  Is there any way to do this kind of OTA upgrade with RAUC using custom handler ? Is RAUC is suitable  for these kind of requirements? does anyone else in  forum came across this scenario, please provide your inputs.

**Comment by ejoerns** [comment]
Hi @saramahe,

RAUC uses the image type (based on the file name extensions of artifacts) together with the *slot type* to figure out, which handler to use for writing the slot update.

For slot type `*.ext4` valid image types are e.g. `*.ext4`, `*.img`, `*.tar.gz`.

`.cpio.gz.u-boot` seems to be an uncommon extension. Where does this originate from?

Maybe have a look into https://github.com/rauc/meta-rauc-community/tree/master/meta-rauc-raspberrypi for a RPI example.

**Reply by saramahe** [reply]
Hi @ejoerns ,
Thanks for the response, 
.cpio.gz.u-boot is the initramfs image extension, unlike .ext4 rootfs image . This initramfs image is the temporary filesystem loaded into memory as an initial ramdisk during early boot to set up the environment. The size of image is lesser compared to .ext4 primary root file system.

This does not require mounting , instead we need to load the initramfs image into RAM like kernel and boot it from there. Once kernel comes up ,  it checks to see if rootfs contains a file "init", and if so it executes it as PID 1. If found, this init process is responsible for bringing the system the rest of the way up, including locating and mounting the real root device (if any).

We use the following to boot the image from u-boot. 
```
load ${BOOT_DEV} ${kernel_addr_r} Image;
load ${BOOT_DEV} ${ramdisk_addr_r}My-dev-image.rootfs.cpio.gz.u-boot;
booti ${kernel_addr_r} ${ramdisk_addr_r} ${fdt_addr}
```

So , I need RAUC to extract the bundles into the inactive partition , skip the installation of rootfs  and transfer the control to u-boot to execute these boot commands from boot.src

BOOT_DEV variable can define from which partition the image to be loaded. 

Can we do this kind of installation using RAUC?
